Cell 1：检查 GPU

In [1]:
import torch
print("GPU 是否可用:", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU 型号:", torch.cuda.get_device_name(0))

GPU 是否可用: True
GPU 型号: Tesla T4


Cell 2：挂载 Google Drive

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Cell 3：验证数据路径

In [4]:
import os
data_path = "/content/drive/MyDrive/SMGC-data/"
print(os.listdir(data_path))

['simulation', 'Mouse_Embryos_S2', 'Human_Lymph_Nodes', 'E18.5_mouse_brain']


Cell 4：克隆 GitHub 代码

In [3]:
!git clone https://huqian122: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@github.com/huqian122/SpaMGCL.git /content/SpaMGCL

Cloning into '/content/SpaMGCL'...
remote: Enumerating objects: 328, done.
remote: Counting objects: 100% (328/328), done.
remote: Compressing objects: 100% (313/313), done.
remote: Total 328 (delta 126), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (328/328), 2.35 MiB | 11.61 MiB/s, done.
Resolving deltas: 100% (126/126), done.


Cell 5：运行 Phase 4 冒烟测试

In [5]:
!cd /content/SpaMGCL/SpaMGCL && python scripts/phase4_smoke.py

/content/SpaMGCL/SpaMGCL/src/models/gcn.py:32: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  tensor = torch.sparse_coo_tensor(indices, values, size=adj.shape).coalesce()
L_rec=0.250091
L_MGCL=3.087323
L_cluster=2.233750
L_spatial=0.000285
WD=[0.33151009678840637, 0.17174333333969116, 0.15729603171348572, 0.13252972066402435]
SC=[0.00014662364264950156, 2.6803680157172494e-05, 4.2885065340669826e-05, 6.826334720244631e-05]
w_WD=[0.2181641012430191, 0.25595822930336, 0.2596829831600189, 0.2661946713924408]
w_SC=[0.21815554797649384, 0.255963534116745, 0.25968626141548157, 0.2661946713924408]
✅ Phase 4 验证通过


Cell 6：安装依赖

In [4]:
%cd /content/SpaMGCL/SpaMGCL
!pip install anndata scanpy

/content/SpaMGCL/SpaMGCL
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.8/188.8 kB 14.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 75.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB 8.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.0/40.0 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 125.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.7/363.7 kB 36.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 130.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.4/69.4 kB 8.3 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.3
    Uninstalling pandas-2.2.3:
      Successfully uninstalled pandas-2.2.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This 

Cell 7：运行 P0（Simulation，基线）

In [7]:
%cd /content/SpaMGCL/SpaMGCL

import os, yaml

# 先拉最新代码
!git pull

# 已存在就跳过
if os.path.exists("results/p0_smoke/metrics.json"):
    print("⏭️ P0 结果已存在，跳过重跑")
else:
    config_path = "configs/p0_smoke.yaml"
    with open(config_path) as f:
        cfg = yaml.safe_load(f)
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)
    cfg["experiment"]["name"] = "p0_smoke"
    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    !python experiments/run_exp.py --config configs/p0_smoke.yaml
    !cat results/p0_smoke/metrics.json

/content/SpaMGCL/SpaMGCL
Already up to date.
⏭️ P0 结果已存在，跳过重跑


Cell 8：跑 HLN-A1

In [8]:
%cd /content/SpaMGCL/SpaMGCL

import os, yaml

if os.path.exists("results/p1_hlna1/metrics.json"):
    print("⏭️ HLN-A1 结果已存在，跳过重跑")
else:
    config_path = "configs/p1_hlna1_base.yaml"
    with open("configs/p0_smoke.yaml") as f:
        cfg = yaml.safe_load(f)

    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["data"]["label_key"] = "Spatial_Label"
    cfg["data"]["n_pca_components"] = 100
    cfg["experiment"]["dataset"] = "HLN-A1"
    cfg["experiment"]["name"] = "p1_hlna1"
    cfg["training"]["epochs"] = 50
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)

    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)

    !python experiments/run_exp.py --config configs/p1_hlna1_base.yaml
    !cat results/p1_hlna1/metrics.json

/content/SpaMGCL/SpaMGCL
⏭️ HLN-A1 结果已存在，跳过重跑


Cell 9：跑 E18.5

In [9]:
%cd /content/SpaMGCL/SpaMGCL

import os, yaml

if os.path.exists("results/p1_e185/metrics.json"):
    print("⏭️ E18.5 结果已存在，跳过重跑")
else:
    config_path = "configs/p1_e185_base.yaml"
    with open("configs/p0_smoke.yaml") as f:
        cfg = yaml.safe_load(f)

    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["data"]["label_key"] = "Combined_Clusters"
    cfg["data"]["n_pca_components"] = 100
    cfg["experiment"]["dataset"] = "E18.5"
    cfg["experiment"]["name"] = "p1_e185"
    cfg["training"]["epochs"] = 50
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)

    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)

    !python experiments/run_exp.py --config configs/p1_e185_base.yaml
    !cat results/p1_e185/metrics.json

/content/SpaMGCL/SpaMGCL
⏭️ E18.5 结果已存在，跳过重跑


Cell 10（三组消融）

In [10]:
%%bash
cd /content/SpaMGCL/SpaMGCL

python - <<'PY'
import yaml

with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base = yaml.safe_load(f)

variants = [
    ("ablation_hlna1_core",   False, False, False),
    ("ablation_hlna1_sc",     True,  True,  False),
    ("ablation_hlna1_sc_snf", True,  True,  True),
]

for name, enabled, sc, snf in variants:
    cfg = yaml.safe_load(yaml.safe_dump(base))

    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)
    cfg["experiment"]["name"] = name
    cfg["loss"]["lambda_cluster"] = 0.1
    cfg["loss"]["lambda_spatial"] = 0.0
    cfg["spatial"] = {
        "enabled": enabled,
        "consistency_weighting": sc,
        "negative_filter": snf,
    }
    cfg["training"]["spatial_start_epoch"] = 11 if enabled else 999

    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"✅ {name}: SC={sc} SNF={snf}")
PY

for name in ablation_hlna1_core ablation_hlna1_sc ablation_hlna1_sc_snf; do
    echo ""
    echo "=========================================="
    echo "🚀 $name"
    echo "=========================================="
    python experiments/run_exp.py --config configs/${name}.yaml || echo "⚠️ $name 失败或已存在"
done

echo ""
echo "🏁 三组消融完成"

✅ ablation_hlna1_core: SC=False SNF=False
✅ ablation_hlna1_sc: SC=True SNF=False
✅ ablation_hlna1_sc_snf: SC=True SNF=True

🚀 ablation_hlna1_core
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial=0.000000 | SC=off | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242313 | mgcl=8.157817 | cluster=2.950293 | spatial=0.000000 | SC=off | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.949508 | spatial=0.000000 | SC=off | SNF=off | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379531 | rec=0.223314 | mgcl=8.156217 | cluster=2.948801 | spatial=0.000000 | SC=off | SNF=off | effC=9.967 | gradC=0.000e+00
epoch 005/050 | total=8.370092 | rec=0.214360 | mgcl=8.155732 | cl

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarn

Cell 11（结果汇总）

In [11]:
import json
import pandas as pd

rows = []
for name in ["ablation_hlna1_core", "ablation_hlna1_sc", "ablation_hlna1_sc_snf"]:
    path = f"/content/SpaMGCL/SpaMGCL/results/{name}/metrics.json"
    try:
        with open(path) as f:
            d = json.load(f)
        rows.append({
            "config": name,
            "SC": d.get("spatial_weighting_enabled"),
            "SNF": d.get("spatial_negative_filter_enabled"),
            "ARI": round(d["ARI"], 4),
            "NMI": round(d["NMI"], 4),
        })
    except FileNotFoundError:
        rows.append({"config": name, "status": "❌ 未完成"})

print(pd.DataFrame(rows).to_string(index=False))

               config    SC   SNF    ARI    NMI
  ablation_hlna1_core False False 0.2894 0.3539
    ablation_hlna1_sc  True False 0.2883 0.3535
ablation_hlna1_sc_snf  True  True 0.2880 0.3529


SC 消融：SC 开启后 ARI 从 0.2894 降到 0.2883（-0.0011），NMI 从 0.3539 降到 0.3535。
SNF 消融：SC+SNF 后 ARI 进一步降到 0.2880，NMI 降到 0.3529。
结论：在当前 fine32 + lambda_cluster=0.1 + lambda_spatial=0 设置下，
      SC 和 SNF 均无正向收益，SNF 甚至进一步轻微下降。
      空间机制的收益问题需要单独设计实验验证，见下一节。

cell12:跑E18.5 数据

In [10]:
%cd /content/SpaMGCL/SpaMGCL

import yaml

with open("configs/e185_ablation_core.yaml") as f:
    base = yaml.safe_load(f)

variants = [
    ("e185_cl03",   32,  0.3),
    ("e185_cl10",   32,  1.0),
    ("e185_cl30",   32,  3.0),
    ("e185_dim64",  64,  1.0),
    ("e185_dim128", 128, 1.0),
]

for name, dim, lam in variants:
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["model"]["fine_dim"] = dim
    cfg["loss"]["lambda_cluster"] = lam
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"✅ {name}: fine_dim={dim}, lambda_cluster={lam}")

/content/SpaMGCL/SpaMGCL
✅ e185_cl03: fine_dim=32, lambda_cluster=0.3
✅ e185_cl10: fine_dim=32, lambda_cluster=1.0
✅ e185_cl30: fine_dim=32, lambda_cluster=3.0
✅ e185_dim64: fine_dim=64, lambda_cluster=1.0
✅ e185_dim128: fine_dim=128, lambda_cluster=1.0


In [11]:
%cd /content/SpaMGCL/SpaMGCL

!python experiments/run_exp.py --config configs/e185_cl03.yaml
!python experiments/run_exp.py --config configs/e185_cl10.yaml
!python experiments/run_exp.py --config configs/e185_cl30.yaml
!python experiments/run_exp.py --config configs/e185_dim64.yaml
!python experiments/run_exp.py --config configs/e185_dim128.yaml

/content/SpaMGCL/SpaMGCL
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=7.864917 | rec=0.198570 | mgcl=7.666347 | cluster=3.303800 | spatial=0.000000 | SC=off | SNF=off | effC=13.909 | gradC=0.000e+00
epoch 002/050 | total=7.853665 | rec=0.188822 | mgcl=7.664843 | cluster=3.302656 | spatial=0.000000 | SC=off | SNF=off | effC=13.922 | gradC=0.000e+0

In [12]:
# 汇总
import json, pandas as pd

rows = []
for name in ["e185_cl03", "e185_cl10", "e185_cl30", "e185_dim64", "e185_dim128"]:
    p = f"results/{name}/metrics.json"
    try:
        with open(p) as f:
            d = json.load(f)
        rows.append({
            "config": name,
            "fine_dim": d["fine_dim"],
            "lambda_cluster": d["loss_history"][-1]["lambda_cluster"],
            "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
            "ARI": round(d["ARI"], 4),
            "NMI": round(d["NMI"], 4),
        })
    except FileNotFoundError:
        rows.append({"config": name, "status": "未完成"})

print("\n===== E18.5 参数扫描结果 =====")
print(pd.DataFrame(rows).to_string(index=False))


===== E18.5 参数扫描结果 =====
     config  fine_dim  lambda_cluster   effC    ARI    NMI
  e185_cl03        32             0.3 13.998 0.2181 0.2004
  e185_cl10        32             1.0 13.993 0.2325 0.2269
  e185_cl30        32             3.0 13.971 0.2120 0.2799
 e185_dim64        64             1.0 13.999 0.2705 0.3727
e185_dim128       128             1.0 13.989 0.2840 0.4334


In [13]:
import json
with open("/content/SpaMGCL/SpaMGCL/results/e185_dim128/metrics.json") as f:
    d = json.load(f)

for k in ["cluster_argmax"]:
    print(k, round(d["candidate_metrics"][k]["ARI"], 4), round(d["candidate_metrics"][k]["NMI"], 4))
for k in ["weighted_views", "mean_views", "global", "concat_views"]:
    print(k, round(d["candidate_metrics"][k]["ARI"], 4), round(d["candidate_metrics"][k]["NMI"], 4))
for k in ["mean_z", "concat_z"]:
    print(k, round(d["representation_metrics"][k]["ARI"], 4), round(d["representation_metrics"][k]["NMI"], 4))
for k in ["rna_spatial", "rna_feature", "atac_spatial", "atac_feature"]:
    print(k, round(d["representation_metrics"]["z_views"][k]["ARI"], 4), round(d["representation_metrics"]["z_views"][k]["NMI"], 4))

cluster_argmax 0.284 0.4334
weighted_views 0.3553 0.5047
mean_views 0.335 0.4898
global 0.3298 0.448
concat_views 0.3298 0.4921
mean_z 0.3096 0.4594
concat_z 0.3375 0.4986
rna_spatial 0.2773 0.4479
rna_feature 0.2252 0.3922
atac_spatial 0.3765 0.4135
atac_feature 0.2828 0.4477


In [18]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_dim128.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_dim256":  {"model": {"fine_dim": 256}},
    "e185_spk5":    {"graphs": {"spatial": {"k": 5}}},
    "e185_ep100":   {"training": {"epochs": 100}, "experiment": {"epochs": 100}},
    "e185_hid64":   {"model": {"gcn_hidden_dim": 64}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_dim256 results/e185_spk5 results/e185_ep100 results/e185_hid64

!python experiments/run_exp.py --config configs/e185_dim256.yaml
!python experiments/run_exp.py --config configs/e185_spk5.yaml
!python experiments/run_exp.py --config configs/e185_ep100.yaml
!python experiments/run_exp.py --config configs/e185_hid64.yaml

rows = [{
    "config": "e185_dim128 (基准)",
    "fine_dim": 128, "spatial_k": 3, "epochs": 50, "hid": 32,
    "ARI": 0.2840, "NMI": 0.4334,
}]
for name in ["e185_dim256", "e185_spk5", "e185_ep100", "e185_hid64"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "fine_dim": d["fine_dim"],
        "spatial_k": d["data"].get("spatial_k", "-"),
        "epochs": d["epochs"],
        "hid": "-",
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_dim256
生成 e185_spk5
生成 e185_ep100
生成 e185_hid64
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=7.832512 | rec=0.154364 | mgcl=7.678148 | cluster=3.296766 | spatial=0.000000 | SC=off | SNF=off | effC=13.993 | gradC=0.000e+00
epoch 002/050 | total=7.793888 | rec=0.135850 | mgcl=7.658038 | cluster=3.296433 | spatial=0.

In [19]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_dim128.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_dim128_mgcl05":  {"loss": {"lambda_mgcl": 0.5}},
    "e185_dim128_mgcl2":   {"loss": {"lambda_mgcl": 2.0}},
    "e185_dim128_t05":     {"model": {"cluster_temperature": 0.5}},
    "e185_dim128_t2":      {"model": {"cluster_temperature": 2.0}},
    "e185_dim128_featk10": {"graphs": {"feature": {"k": 10}}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_dim128_mgcl05 results/e185_dim128_mgcl2 \
       results/e185_dim128_t05 results/e185_dim128_t2 results/e185_dim128_featk10

!python experiments/run_exp.py --config configs/e185_dim128_mgcl05.yaml
!python experiments/run_exp.py --config configs/e185_dim128_mgcl2.yaml
!python experiments/run_exp.py --config configs/e185_dim128_t05.yaml
!python experiments/run_exp.py --config configs/e185_dim128_t2.yaml
!python experiments/run_exp.py --config configs/e185_dim128_featk10.yaml

rows = [{"config": "e185_dim128 (基准)", "ARI": 0.2840, "NMI": 0.4334}]
for name in ["e185_dim128_mgcl05", "e185_dim128_mgcl2",
             "e185_dim128_t05", "e185_dim128_t2", "e185_dim128_featk10"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_dim128_mgcl05
生成 e185_dim128_mgcl2
生成 e185_dim128_t05
生成 e185_dim128_t2
生成 e185_dim128_featk10
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=3.995173 | rec=0.160812 | mgcl=7.668722 | cluster=3.298100 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=3.976383 | rec=0.146324 

In [20]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_dim128_t05.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_t05_mgcl2":       {"loss": {"lambda_mgcl": 2.0}},
    "e185_t05_featk10":     {"graphs": {"feature": {"k": 10}}},
    "e185_t05_mgcl2_featk10": {"loss": {"lambda_mgcl": 2.0}, "graphs": {"feature": {"k": 10}}},
    "e185_t03":             {"model": {"cluster_temperature": 0.3}},
    "e185_t01":             {"model": {"cluster_temperature": 0.1}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_t05_mgcl2 results/e185_t05_featk10 \
       results/e185_t05_mgcl2_featk10 results/e185_t03 results/e185_t01

!python experiments/run_exp.py --config configs/e185_t05_mgcl2.yaml
!python experiments/run_exp.py --config configs/e185_t05_featk10.yaml
!python experiments/run_exp.py --config configs/e185_t05_mgcl2_featk10.yaml
!python experiments/run_exp.py --config configs/e185_t03.yaml
!python experiments/run_exp.py --config configs/e185_t01.yaml

rows = [{"config": "e185_dim128_t05 (当前最佳)", "ARI": 0.4071, "NMI": 0.4401}]
for name in ["e185_t05_mgcl2", "e185_t05_featk10", "e185_t05_mgcl2_featk10",
             "e185_t03", "e185_t01"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_t05_mgcl2
生成 e185_t05_featk10
生成 e185_t05_mgcl2_featk10
生成 e185_t03
生成 e185_t01
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=15.498254 | rec=0.160812 | mgcl=7.668721 | cluster=3.298072 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=15.466586 | rec=0.147807 | mgcl=7.6593

In [21]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_t05_mgcl2.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_t04_mgcl2":   {"model": {"cluster_temperature": 0.4}},
    "e185_t03_mgcl2":   {"model": {"cluster_temperature": 0.3}},
    "e185_t05_mgcl3":   {"loss": {"lambda_mgcl": 3.0}},
    "e185_t05_mgcl2_fk15": {"graphs": {"feature": {"k": 15}}},
    "e185_t03_featk10": {"model": {"cluster_temperature": 0.3},
                         "graphs": {"feature": {"k": 10}}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_t04_mgcl2 results/e185_t03_mgcl2 results/e185_t05_mgcl3 \
       results/e185_t05_mgcl2_fk15 results/e185_t03_featk10

!python experiments/run_exp.py --config configs/e185_t04_mgcl2.yaml
!python experiments/run_exp.py --config configs/e185_t03_mgcl2.yaml
!python experiments/run_exp.py --config configs/e185_t05_mgcl3.yaml
!python experiments/run_exp.py --config configs/e185_t05_mgcl2_fk15.yaml
!python experiments/run_exp.py --config configs/e185_t03_featk10.yaml

rows = [{"config": "e185_t05_mgcl2 (当前最佳)", "ARI": 0.4559, "NMI": 0.4410}]
for name in ["e185_t04_mgcl2", "e185_t03_mgcl2", "e185_t05_mgcl3",
             "e185_t05_mgcl2_fk15", "e185_t03_featk10"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_t04_mgcl2
生成 e185_t03_mgcl2
生成 e185_t05_mgcl3
生成 e185_t05_mgcl2_fk15
生成 e185_t03_featk10
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=15.498254 | rec=0.160812 | mgcl=7.668721 | cluster=3.298058 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=15.466587 | rec=0.147807 | mg

In [22]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_t05_mgcl3.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_t05_mgcl5":    {"loss": {"lambda_mgcl": 5.0}},
    "e185_t05_mgcl4":    {"loss": {"lambda_mgcl": 4.0}},
    "e185_t03_mgcl3":    {"model": {"cluster_temperature": 0.3}, "loss": {"lambda_mgcl": 3.0}},
    "e185_t04_mgcl3":    {"model": {"cluster_temperature": 0.4}, "loss": {"lambda_mgcl": 3.0}},
    "e185_t035_mgcl3":   {"model": {"cluster_temperature": 0.35}, "loss": {"lambda_mgcl": 3.0}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_t05_mgcl5 results/e185_t05_mgcl4 results/e185_t03_mgcl3 \
       results/e185_t04_mgcl3 results/e185_t035_mgcl3

!python experiments/run_exp.py --config configs/e185_t05_mgcl5.yaml
!python experiments/run_exp.py --config configs/e185_t05_mgcl4.yaml
!python experiments/run_exp.py --config configs/e185_t03_mgcl3.yaml
!python experiments/run_exp.py --config configs/e185_t04_mgcl3.yaml
!python experiments/run_exp.py --config configs/e185_t035_mgcl3.yaml

rows = [
    {"config": "e185_t05_mgcl3", "ARI": 0.4768, "NMI": 0.4454},
    {"config": "e185_t03_mgcl2", "ARI": 0.4625, "NMI": 0.4679},
]
for name in ["e185_t05_mgcl5", "e185_t05_mgcl4", "e185_t03_mgcl3",
             "e185_t04_mgcl3", "e185_t035_mgcl3"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_t05_mgcl5
生成 e185_t05_mgcl4
生成 e185_t03_mgcl3
生成 e185_t04_mgcl3
生成 e185_t035_mgcl3
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=38.504417 | rec=0.160812 | mgcl=7.668721 | cluster=3.298072 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=38.444881 | rec=0.149041 | mgcl=7.6

In [23]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_t03_mgcl3.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "e185_t03_mgcl4":   {"loss": {"lambda_mgcl": 4.0}},
    "e185_t025_mgcl3":  {"model": {"cluster_temperature": 0.25}},
    "e185_t02_mgcl3":   {"model": {"cluster_temperature": 0.2}},
    "e185_t025_mgcl4":  {"model": {"cluster_temperature": 0.25},
                         "loss": {"lambda_mgcl": 4.0}},
    "e185_dim256_t03_mgcl3": {"model": {"fine_dim": 256}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/e185_t03_mgcl4 results/e185_t025_mgcl3 results/e185_t02_mgcl3 \
       results/e185_t025_mgcl4 results/e185_dim256_t03_mgcl3

!python experiments/run_exp.py --config configs/e185_t03_mgcl4.yaml
!python experiments/run_exp.py --config configs/e185_t025_mgcl3.yaml
!python experiments/run_exp.py --config configs/e185_t02_mgcl3.yaml
!python experiments/run_exp.py --config configs/e185_t025_mgcl4.yaml
!python experiments/run_exp.py --config configs/e185_dim256_t03_mgcl3.yaml

rows = [{"config": "e185_t03_mgcl3 (当前最佳)", "ARI": 0.4963, "NMI": 0.4712}]
for name in ["e185_t03_mgcl4", "e185_t025_mgcl3", "e185_t02_mgcl3",
             "e185_t025_mgcl4", "e185_dim256_t03_mgcl3"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 e185_t03_mgcl4
生成 e185_t025_mgcl3
生成 e185_t02_mgcl3
生成 e185_t025_mgcl4
生成 e185_dim256_t03_mgcl3
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=30.835695 | rec=0.160812 | mgcl=7.668721 | cluster=3.298035 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=30.785578 | rec=0.148728 | 

论文里建议主报 t03_mgcl3（ARI 高、NMI 也不差），附 t025_mgcl4 作为平衡点

In [24]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/e185_t03_mgcl3.yaml") as f:
    e185_best = yaml.safe_load(f)

# 派生 HLN-A1 版本
cfg = yaml.safe_load(yaml.safe_dump(e185_best))
cfg["experiment"]["name"] = "hlna1_t03_mgcl3"
cfg["experiment"]["dataset"] = "HLN-A1"
cfg["data"]["label_key"] = "Spatial_Label"
cfg["model"]["num_clusters"] = 10
cfg["clustering"]["n_clusters"] = 10
cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
cfg["output"] = {"root": "results"}
with open("configs/hlna1_t03_mgcl3.yaml", "w") as f:
    yaml.dump(cfg, f, default_flow_style=False)
print("生成 hlna1_t03_mgcl3")

!rm -rf results/hlna1_t03_mgcl3
!python experiments/run_exp.py --config configs/hlna1_t03_mgcl3.yaml

with open("results/hlna1_t03_mgcl3/metrics.json") as f:
    d = json.load(f)
print(f"HLN-A1 t03_mgcl3: ARI={d['ARI']:.4f}  NMI={d['NMI']:.4f}")
print(f"HLN-A1 基准:      ARI=0.2894  NMI=0.3539")

/content/SpaMGCL/SpaMGCL
生成 hlna1_t03_mgcl3
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
Dataset: HLN-A

| 数据集 | fine_dim | cluster_temp | lambda_mgcl | ARI | NMI |
|---|---|---|---|---|---|
| HLN-A1 | 32 | 1.0 | 1.0 | 0.2894 | 0.3539 |
| E18.5 | 128 | 0.3 | 3.0 | 0.4963 | 0.4712 |

In [6]:
!ls /content/SpaMGCL/SpaMGCL/configs/ | grep -i hlna1
!ls /content/SpaMGCL/SpaMGCL/configs/ | grep -i e185

diagnostic_hlna1_fine128.yaml
diagnostic_hlna1_fine16.yaml
diagnostic_hlna1_fine32.yaml
diagnostic_hlna1_fine64.yaml
p1_hlna1.yaml
e185_ablation_core.yaml
e185_ablation_sc_snf.yaml
e185_ablation_sc.yaml
p1_e185.yaml


In [7]:
%cd /content/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

# 用 fine32 作为 base
with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base = yaml.safe_load(f)

# 先把 base 对齐到 HLN-A1 最佳 baseline
base["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
base["data"]["label_key"] = "Spatial_Label"
base["loss"]["lambda_cluster"] = 0.1
base["loss"]["lambda_spatial"] = 0.0
base["spatial"] = {"enabled": False, "consistency_weighting": False, "negative_filter": False}
base["training"]["cluster_head_lr_multiplier"] = 8.0
base["training"]["spatial_start_epoch"] = 999
base["output"] = {"root": "results"}

variants = {
    "hlna1_t2":        {"model": {"cluster_temperature": 2.0}},
    "hlna1_t05":       {"model": {"cluster_temperature": 0.5}},
    "hlna1_mgcl2":     {"loss": {"lambda_mgcl": 2.0}},
    "hlna1_mgcl3":     {"loss": {"lambda_mgcl": 3.0}},
    "hlna1_t2_mgcl2":  {"model": {"cluster_temperature": 2.0}, "loss": {"lambda_mgcl": 2.0}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/hlna1_t2 results/hlna1_t05 results/hlna1_mgcl2 results/hlna1_mgcl3 results/hlna1_t2_mgcl2

!python experiments/run_exp.py --config configs/hlna1_t2.yaml
!python experiments/run_exp.py --config configs/hlna1_t05.yaml
!python experiments/run_exp.py --config configs/hlna1_mgcl2.yaml
!python experiments/run_exp.py --config configs/hlna1_mgcl3.yaml
!python experiments/run_exp.py --config configs/hlna1_t2_mgcl2.yaml

rows = [{"config": "hlna1 基准 (fine32, λc=0.1, t=1, mgcl=1)", "ARI": 0.2894, "NMI": 0.3539}]
for name in ["hlna1_t2", "hlna1_t05", "hlna1_mgcl2", "hlna1_mgcl3", "hlna1_t2_mgcl2"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/content/SpaMGCL/SpaMGCL
生成 hlna1_t2
生成 hlna1_t05
生成 hlna1_mgcl2
生成 hlna1_mgcl3
生成 hlna1_t2_mgcl2
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:149: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:

| 数据集 | fine_dim | cluster_temp | lambda_mgcl | ARI | NMI |
|---|---|---|---|---|---|
| HLN-A1 | 32 | 1.0 | 1.0 | 0.2894 | 0.3539 |
| E18.5 | 128 | 0.3 | 3.0 | 0.4963 | 0.4712 |

cella:修正诊断配置的 data.root

In [11]:
import os
import yaml

os.chdir("/content/SpaMGCL/SpaMGCL")

# 修正所有诊断配置文件的 data.root
for d in [16, 32, 64, 128]:
    path = f"configs/diagnostic_hlna1_fine{d}.yaml"
    with open(path, "r") as f:
        cfg = yaml.safe_load(f)
    cfg["data"]["root"] = "/content/drive/MyDrive/SMGC-data/"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"✅ 已修正 {path} 的 data.root")

# 验证一下
for d in [16, 32, 64, 128]:
    with open(f"configs/diagnostic_hlna1_fine{d}.yaml", "r") as f:
        cfg = yaml.safe_load(f)
    print(f"fine{d}: data.root = {cfg['data']['root']}")

✅ 已修正 configs/diagnostic_hlna1_fine16.yaml 的 data.root
✅ 已修正 configs/diagnostic_hlna1_fine32.yaml 的 data.root
✅ 已修正 configs/diagnostic_hlna1_fine64.yaml 的 data.root
✅ 已修正 configs/diagnostic_hlna1_fine128.yaml 的 data.root
fine16: data.root = /content/drive/MyDrive/SMGC-data/
fine32: data.root = /content/drive/MyDrive/SMGC-data/
fine64: data.root = /content/drive/MyDrive/SMGC-data/
fine128: data.root = /content/drive/MyDrive/SMGC-data/


cellb:批量运行 fine_dim=16/32/64/128

In [12]:
%%bash
cd /content/SpaMGCL/SpaMGCL

for d in 16 32 64 128; do
    echo "=========================================="
    echo "🚀 开始 fine_dim=${d}"
    echo "=========================================="
    python experiments/run_exp.py --config configs/diagnostic_hlna1_fine${d}.yaml
    echo "✅ fine_dim=${d} 完成"
    echo ""
done

🚀 开始 fine_dim=16
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.405149 | rec=0.248008 | mgcl=8.157142 | cluster=2.959329 | spatial=0.000000 | SC=off | SNF=off | effC=9.858 | gradC=0.000e+00
epoch 002/050 | total=8.394496 | rec=0.237817 | mgcl=8.156678 | cluster=2.958024 | spatial=0.000000 | SC=off | SNF=off | effC=9.871 | gradC=0.000e+00
epoch 003/050 | total=8.384409 | rec=0.228064 | mgcl=8.156344 | cluster=2.956799 | spatial=0.000000 | SC=off | SNF=off | effC=9.882 | gradC=0.000e+00
epoch 004/050 | total=8.374840 | rec=0.218736 | mgcl=8.156104 | cluster=2.955626 | spatial=0.000000 | SC=off | SNF=off | effC=9.894 | gradC=0.000e+00
epoch 005/050 | total=8.365738 | rec=0.209811 | mgcl=8.155927 | cluster=2.954504 | spatial=0.000000 | SC=off | SNF=off | effC=9.904 | gradC=0.000e+00
epoch 006/050 | total=8.357061 | rec=0.201272

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:101: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarn

cellc:汇总 fine_dim 主指标

In [13]:
import json
import pandas as pd

rows = []
for d in [16, 32, 64, 128]:
    path = f"/content/SpaMGCL/SpaMGCL/results/diagnostic_hlna1_fine{d}/metrics.json"
    try:
        with open(path, "r") as f:
            data = json.load(f)
        rows.append({
            "fine_dim": d,
            "Z_mean_ARI": data.get("z_mean_ari", None),
            "Z_concat_ARI": data.get("z_concat_ari", None),
            "Q_argmax_ARI": data.get("ARI", None),
            "NMI": data.get("NMI", None)
        })
    except FileNotFoundError:
        rows.append({"fine_dim": d, "status": "❌ 未找到"})

df = pd.DataFrame(rows)
print(df.to_string(index=False))

 fine_dim Z_mean_ARI Z_concat_ARI  Q_argmax_ARI      NMI
       16       None         None      0.136410 0.193117
       32       None         None      0.262954 0.341379
       64       None         None      0.225507 0.303100
      128       None         None      0.203812 0.306323


| fine_dim |        ARI |        NMI | 排名     |
| -------: | ---------: | ---------: | ------ |
|       16 |     0.1365 |     0.1932 | 最差     |
|   **32** | **0.2629** | **0.3408** | **最好** |
|       64 |     0.2254 |     0.3030 | 次优     |
|      128 |     0.2045 |     0.3066 | 再下降    |


celld:查看 fine_dim 的表示层指标

读取四个 fine_dim 的 representation_metrics

In [14]:
import json

path = "results/diagnostic_hlna1_fine32/metrics.json"

with open(path, "r") as f:
    m = json.load(f)

print("顶层 keys:")
print(m.keys())

print("\n完整 metrics:")
print(json.dumps(m, indent=2, ensure_ascii=False))

顶层 keys:
dict_keys(['dataset', 'seed', 'device', 'epochs', 'warm_up_epochs', 'spatial_start_epoch', 'spatial_mechanism_enabled', 'spatial_weighting_enabled', 'spatial_negative_filter_enabled', 'cluster_enabled_epochs', 'spatial_enabled_epochs', 'clustering_method_requested', 'clustering_method_used', 'clustering_embedding', 'candidate_metrics', 'representation_metrics', 'fine_dim', 'num_clusters', 'ARI', 'NMI', 'metrics', 'final_weights', 'final_cluster_diagnostics', 'data', 'graph_stats', 'loss_history'])

完整 metrics:
{
  "dataset": "HLN-A1",
  "seed": 0,
  "device": "cuda",
  "epochs": 50,
  "warm_up_epochs": 10,
  "spatial_start_epoch": 51,
  "spatial_mechanism_enabled": false,
  "spatial_weighting_enabled": false,
  "spatial_negative_filter_enabled": false,
  "cluster_enabled_epochs": [
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35

celle：对比 mean_z、concat_z 与 Q_argmax

生成并运行两个微调实验：• diag_fine32_lr2x：cluster_head_lr_multiplier=2.0•

diag_fine32_cl1：lambda_cluster=1.0

In [15]:
%%bash
cd /content/SpaMGCL/SpaMGCL

# 复制 fine32 配置，创建两个变体
python - <<'PY'
import yaml

with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base = yaml.safe_load(f)

# 实验 A：降低 cluster_head_lr_multiplier 到 2x
cfg_a = yaml.safe_load(yaml.safe_dump(base))
cfg_a["experiment"]["name"] = "diag_fine32_lr2x"
cfg_a["training"]["cluster_head_lr_multiplier"] = 2.0
cfg_a["output"]["dir"] = "results/diag_fine32_lr2x/"
with open("configs/diag_fine32_lr2x.yaml", "w") as f:
    yaml.dump(cfg_a, f, default_flow_style=False)

# 实验 B：降低 lambda_cluster 到 1.0
cfg_b = yaml.safe_load(yaml.safe_dump(base))
cfg_b["experiment"]["name"] = "diag_fine32_cl1"
cfg_b["loss"]["lambda_cluster"] = 1.0
cfg_b["output"]["dir"] = "results/diag_fine32_cl1/"
with open("configs/diag_fine32_cl1.yaml", "w") as f:
    yaml.dump(cfg_b, f, default_flow_style=False)

print("✅ 生成 diag_fine32_lr2x.yaml 和 diag_fine32_cl1.yaml")
PY

echo "=========================================="
echo "🚀 实验 A: fine_dim=32, cluster_head_lr_multiplier=2.0"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine32_lr2x.yaml

echo ""
echo "=========================================="
echo "🚀 实验 B: fine_dim=32, lambda_cluster=1.0"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine32_cl1.yaml

echo ""
echo "🏁 两个微调实验完成"

✅ 生成 diag_fine32_lr2x.yaml 和 diag_fine32_cl1.yaml
🚀 实验 A: fine_dim=32, cluster_head_lr_multiplier=2.0
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.002 (2x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial=0.000000 | SC=off | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242313 | mgcl=8.157817 | cluster=2.950877 | spatial=0.000000 | SC=off | SNF=off | effC=9.952 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.950612 | spatial=0.000000 | SC=off | SNF=off | effC=9.954 | gradC=0.000e+00
epoch 004/050 | total=8.379531 | rec=0.223314 | mgcl=8.156217 | cluster=2.950364 | spatial=0.000000 | SC=off | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 005/050 | total=8.370091 | rec=0.214360 | mgcl=8.155731 | cluster=2.950131 | spatial=0.000000 | SC=off |

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:101: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarn

cellf：读取 fine_dim 表示层指标表。

目的：读取四个 fine_dim 的 metrics.json，从 representation_metrics 和 candidate_metrics 里提取：

In [16]:
import os
import json
import pandas as pd

fine_dims = [16, 32, 64, 128]
rows = []

for d in fine_dims:
    path = f"results/diagnostic_hlna1_fine{d}/metrics.json"

    with open(path, "r") as f:
        m = json.load(f)

    rep = m["representation_metrics"]
    cand = m["candidate_metrics"]

    rows.append({
        "fine_dim": d,

        "RNA_spatial_Z_ARI": rep["z_views"]["rna_spatial"]["ARI"],
        "RNA_feature_Z_ARI": rep["z_views"]["rna_feature"]["ARI"],
        "ADT_spatial_Z_ARI": rep["z_views"]["adt_spatial"]["ARI"],
        "ADT_feature_Z_ARI": rep["z_views"]["adt_feature"]["ARI"],

        "Z_mean_ARI": rep["mean_z"]["ARI"],
        "Z_concat_ARI": rep["concat_z"]["ARI"],

        "Q_argmax_ARI": cand["cluster_argmax"]["ARI"],
        "Q_argmax_NMI": cand["cluster_argmax"]["NMI"],
    })

df = pd.DataFrame(rows)
display(df)

,fine_dim,RNA_spatial_Z_ARI,RNA_feature_Z_ARI,ADT_spatial_Z_ARI,ADT_feature_Z_ARI,Z_mean_ARI,Z_concat_ARI,Q_argmax_ARI,Q_argmax_NMI
0,16,0.142404,0.122812,0.147725,0.121369,0.173640,0.154628,0.136410,0.193117
1,32,0.262079,0.233164,0.189612,0.159578,0.238379,0.271365,0.280820,0.351580
2,64,0.207654,0.178527,0.212820,0.182648,0.212208,0.238012,0.225507,0.303100
3,128,0.263503,0.203931,0.247616,0.205633,0.276440,0.279341,0.203812,0.306323


cellg：meanZ / concatZ / Q 对比表

In [17]:
rows = []

for d in [16, 32, 64, 128]:
    with open(f"results/diagnostic_hlna1_fine{d}/metrics.json") as f:
        m = json.load(f)

    rep = m["representation_metrics"]
    q = m["candidate_metrics"]["cluster_argmax"]

    rows.append({
        "fine_dim": d,
        "meanZ_ARI": rep["mean_z"]["ARI"],
        "meanZ_NMI": rep["mean_z"]["NMI"],
        "concatZ_ARI": rep["concat_z"]["ARI"],
        "concatZ_NMI": rep["concat_z"]["NMI"],
        "Q_ARI": q["ARI"],
        "Q_NMI": q["NMI"],
    })

pd.DataFrame(rows)

,fine_dim,meanZ_ARI,meanZ_NMI,concatZ_ARI,concatZ_NMI,Q_ARI,Q_NMI
0,16,0.173640,0.242702,0.154628,0.225108,0.136410,0.193117
1,32,0.238379,0.334134,0.271365,0.360825,0.280820,0.351580
2,64,0.212208,0.300818,0.238012,0.334844,0.225507,0.303100
3,128,0.276440,0.362924,0.279341,0.372164,0.203812,0.306323


cellh：lambda_cluster 变体批量实验（cl05 / cl01 / cl2 / fine64_cl1）

作用：生成并运行四个变体：• diag_fine32_cl05：fine32，λ=0.5

• diag_fine32_cl01：fine32，λ=0.1

• diag_fine32_cl20：fine32，λ=2.0（配置文件名是 cl20）

• diag_fine64_cl1：fine64，λ=1.0

In [18]:
%%bash
cd /content/SpaMGCL/SpaMGCL

python - <<'PY'
import yaml

# 读 fine32 基准配置
with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base32 = yaml.safe_load(f)

# 读 fine64 基准配置
with open("configs/diagnostic_hlna1_fine64.yaml") as f:
    base64 = yaml.safe_load(f)

# 三个 lambda 变体（fine32）
for lam in [0.5, 0.1, 2.0]:
    cfg = yaml.safe_load(yaml.safe_dump(base32))
    cfg["experiment"]["name"] = f"diag_fine32_cl{str(lam).replace('.','')}"
    cfg["loss"]["lambda_cluster"] = lam
    cfg["output"]["dir"] = f"results/diag_fine32_cl{str(lam).replace('.','')}/"
    fname = f"configs/diag_fine32_cl{str(lam).replace('.','')}.yaml"
    with open(fname, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"✅ 生成 {fname}")

# fine64 + lambda_cluster=1
cfg = yaml.safe_load(yaml.safe_dump(base64))
cfg["experiment"]["name"] = "diag_fine64_cl1"
cfg["loss"]["lambda_cluster"] = 1.0
cfg["output"]["dir"] = "results/diag_fine64_cl1/"
with open("configs/diag_fine64_cl1.yaml", "w") as f:
    yaml.dump(cfg, f, default_flow_style=False)
print("✅ 生成 configs/diag_fine64_cl1.yaml")
PY

echo ""
echo "=========================================="
echo "🚀 实验 C: fine32, λ_cluster=0.5"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine32_cl05.yaml

echo ""
echo "=========================================="
echo "🚀 实验 D: fine32, λ_cluster=0.1"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine32_cl01.yaml

echo ""
echo "=========================================="
echo "🚀 实验 E: fine32, λ_cluster=2.0"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine32_cl2.yaml

echo ""
echo "=========================================="
echo "🚀 实验 F: fine64, λ_cluster=1.0"
echo "=========================================="
python experiments/run_exp.py --config configs/diag_fine64_cl1.yaml

echo ""
echo "🏁 全部 4 个实验完成"
import json
import pandas as pd

rows = []
for name in ["diag_fine32_cl05", "diag_fine32_cl01", "diag_fine32_cl2", "diag_fine64_cl1"]:
    path = f"/content/SpaMGCL/SpaMGCL/results/{name}/metrics.json"
    try:
        with open(path) as f:
            data = json.load(f)
        rows.append({
            "实验": name,
            "λ_cluster": data["loss_history"][-1]["lambda_cluster"],
            "ARI": round(data["ARI"], 4),
            "NMI": round(data["NMI"], 4),
        })
    except FileNotFoundError:
        rows.append({"实验": name, "状态": "❌ 未找到"})

df = pd.DataFrame(rows)
print(df.to_string(index=False))

✅ 生成 configs/diag_fine32_cl05.yaml
✅ 生成 configs/diag_fine32_cl01.yaml
✅ 生成 configs/diag_fine32_cl20.yaml
✅ 生成 configs/diag_fine64_cl1.yaml

🚀 实验 C: fine32, λ_cluster=0.5
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial=0.000000 | SC=off | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242313 | mgcl=8.157817 | cluster=2.950293 | spatial=0.000000 | SC=off | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.949508 | spatial=0.000000 | SC=off | SNF=off | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379530 | rec=0.223314 | mgcl=8.156216 | cluster=2.948801 | spatial=0.000000 | SC=off | SNF=off | effC=9.967 | gradC=0.000e+00
epoch 005/050 | total=8.370092 | rec=0.214

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:101: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarn

CalledProcessError: Command 'b'cd /content/SpaMGCL/SpaMGCL\n\npython - <<\'PY\'\nimport yaml\n\n# \xe8\xaf\xbb fine32 \xe5\x9f\xba\xe5\x87\x86\xe9\x85\x8d\xe7\xbd\xae\nwith open("configs/diagnostic_hlna1_fine32.yaml") as f:\n    base32 = yaml.safe_load(f)\n\n# \xe8\xaf\xbb fine64 \xe5\x9f\xba\xe5\x87\x86\xe9\x85\x8d\xe7\xbd\xae\nwith open("configs/diagnostic_hlna1_fine64.yaml") as f:\n    base64 = yaml.safe_load(f)\n\n# \xe4\xb8\x89\xe4\xb8\xaa lambda \xe5\x8f\x98\xe4\xbd\x93\xef\xbc\x88fine32\xef\xbc\x89\nfor lam in [0.5, 0.1, 2.0]:\n    cfg = yaml.safe_load(yaml.safe_dump(base32))\n    cfg["experiment"]["name"] = f"diag_fine32_cl{str(lam).replace(\'.\',\'\')}"\n    cfg["loss"]["lambda_cluster"] = lam\n    cfg["output"]["dir"] = f"results/diag_fine32_cl{str(lam).replace(\'.\',\'\')}/"\n    fname = f"configs/diag_fine32_cl{str(lam).replace(\'.\',\'\')}.yaml"\n    with open(fname, "w") as f:\n        yaml.dump(cfg, f, default_flow_style=False)\n    print(f"\xe2\x9c\x85 \xe7\x94\x9f\xe6\x88\x90 {fname}")\n\n# fine64 + lambda_cluster=1\ncfg = yaml.safe_load(yaml.safe_dump(base64))\ncfg["experiment"]["name"] = "diag_fine64_cl1"\ncfg["loss"]["lambda_cluster"] = 1.0\ncfg["output"]["dir"] = "results/diag_fine64_cl1/"\nwith open("configs/diag_fine64_cl1.yaml", "w") as f:\n    yaml.dump(cfg, f, default_flow_style=False)\nprint("\xe2\x9c\x85 \xe7\x94\x9f\xe6\x88\x90 configs/diag_fine64_cl1.yaml")\nPY\n\necho ""\necho "=========================================="\necho "\xf0\x9f\x9a\x80 \xe5\xae\x9e\xe9\xaa\x8c C: fine32, \xce\xbb_cluster=0.5"\necho "=========================================="\npython experiments/run_exp.py --config configs/diag_fine32_cl05.yaml\n\necho ""\necho "=========================================="\necho "\xf0\x9f\x9a\x80 \xe5\xae\x9e\xe9\xaa\x8c D: fine32, \xce\xbb_cluster=0.1"\necho "=========================================="\npython experiments/run_exp.py --config configs/diag_fine32_cl01.yaml\n\necho ""\necho "=========================================="\necho "\xf0\x9f\x9a\x80 \xe5\xae\x9e\xe9\xaa\x8c E: fine32, \xce\xbb_cluster=2.0"\necho "=========================================="\npython experiments/run_exp.py --config configs/diag_fine32_cl2.yaml\n\necho ""\necho "=========================================="\necho "\xf0\x9f\x9a\x80 \xe5\xae\x9e\xe9\xaa\x8c F: fine64, \xce\xbb_cluster=1.0"\necho "=========================================="\npython experiments/run_exp.py --config configs/diag_fine64_cl1.yaml\n\necho ""\necho "\xf0\x9f\x8f\x81 \xe5\x85\xa8\xe9\x83\xa8 4 \xe4\xb8\xaa\xe5\xae\x9e\xe9\xaa\x8c\xe5\xae\x8c\xe6\x88\x90"\nimport json\nimport pandas as pd\n\nrows = []\nfor name in ["diag_fine32_cl05", "diag_fine32_cl01", "diag_fine32_cl2", "diag_fine64_cl1"]:\n    path = f"/content/SpaMGCL/SpaMGCL/results/{name}/metrics.json"\n    try:\n        with open(path) as f:\n            data = json.load(f)\n        rows.append({\n            "\xe5\xae\x9e\xe9\xaa\x8c": name,\n            "\xce\xbb_cluster": data["loss_history"][-1]["lambda_cluster"],\n            "ARI": round(data["ARI"], 4),\n            "NMI": round(data["NMI"], 4),\n        })\n    except FileNotFoundError:\n        rows.append({"\xe5\xae\x9e\xe9\xaa\x8c": name, "\xe7\x8a\xb6\xe6\x80\x81": "\xe2\x9d\x8c \xe6\x9c\xaa\xe6\x89\xbe\xe5\x88\xb0"})\n\ndf = pd.DataFrame(rows)\nprint(df.to_string(index=False))\n'' returned non-zero exit status 2.

| 设置                        |        ARI |        NMI |
| ------------------------- | ---------: | ---------: |
| fine32, λ_cluster=5（之前）   |     0.2808 |     0.3516 |
| fine32, λ_cluster=0.5     |     0.2840 |     0.3523 |
| **fine32, λ_cluster=0.1** | **0.2894** | **0.3539** |
| fine64, λ_cluster=1       |     0.2123 |     0.3064 |


cell I：lambda_cluster 结果表格

In [19]:
%%bash
cd /content/SpaMGCL/SpaMGCL
python - <<'PY'
import yaml
with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base = yaml.safe_load(f)
cfg = yaml.safe_load(yaml.safe_dump(base))
cfg["experiment"]["name"] = "diag_fine32_cl0"
cfg["loss"]["lambda_cluster"] = 0.0
cfg["output"]["dir"] = "results/diag_fine32_cl0/"
with open("configs/diag_fine32_cl0.yaml", "w") as f:
    yaml.dump(cfg, f, default_flow_style=False)
print("✅ 生成 diag_fine32_cl0.yaml")
PY

python experiments/run_exp.py --config configs/diag_fine32_cl0.yaml

✅ 生成 diag_fine32_cl0.yaml
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial=0.000000 | SC=off | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242313 | mgcl=8.157817 | cluster=2.950293 | spatial=0.000000 | SC=off | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.949508 | spatial=0.000000 | SC=off | SNF=off | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379531 | rec=0.223314 | mgcl=8.156217 | cluster=2.948801 | spatial=0.000000 | SC=off | SNF=off | effC=9.967 | gradC=0.000e+00
epoch 005/050 | total=8.370092 | rec=0.214360 | mgcl=8.155732 | cluster=2.948162 | spatial=0.000000 | SC=off | SNF=off | effC=9.972 | gradC=0.000e+00
epoch 006/050 | total=8.361096 | rec

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:101: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(


| λ_cluster |        ARI |        NMI |
| --------: | ---------: | ---------: |
|         0 |     0.0496 |     0.0969 |
|   **0.1** | **0.2894** | **0.3539** |
|       0.5 |     0.2840 |     0.3523 |
|         5 |     0.2808 |     0.3516 |



cell j：Spatial 空间机制消融实验：diag_fine32_spatial，λ_cluster=0

In [20]:
%%bash
cd /content/SpaMGCL/SpaMGCL
python - <<'PY'
import yaml
with open("configs/diagnostic_hlna1_fine32.yaml") as f:
    base = yaml.safe_load(f)
cfg = yaml.safe_load(yaml.safe_dump(base))
cfg["experiment"]["name"] = "diag_fine32_spatial"
cfg["loss"]["lambda_cluster"] = 1.0
cfg["loss"]["lambda_spatial"] = 0.1
cfg["spatial"] = {
    "enabled": True,
    "consistency_weighting": True,
    "negative_filter": False   # 先只加 SC，SNF 保持关闭
}
cfg["training"]["spatial_start_epoch"] = 11
cfg["output"]["dir"] = "results/diag_fine32_spatial/"
with open("configs/diag_fine32_spatial.yaml", "w") as f:
    yaml.dump(cfg, f, default_flow_style=False)
print("✅ 生成 diag_fine32_spatial.yaml")
PY

python experiments/run_exp.py --config configs/diag_fine32_spatial.yaml

✅ 生成 diag_fine32_spatial.yaml
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411675 | rec=0.252383 | mgcl=8.159292 | cluster=2.951158 | spatial=0.000000 | SC=on | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242313 | mgcl=8.157817 | cluster=2.950293 | spatial=0.000000 | SC=on | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389495 | rec=0.232628 | mgcl=8.156867 | cluster=2.949508 | spatial=0.000000 | SC=on | SNF=off | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379530 | rec=0.223314 | mgcl=8.156216 | cluster=2.948801 | spatial=0.000000 | SC=on | SNF=off | effC=9.967 | gradC=0.000e+00
epoch 005/050 | total=8.370091 | rec=0.214360 | mgcl=8.155732 | cluster=2.948162 | spatial=0.000000 | SC=on | SNF=off | effC=9.972 | gradC=0.000e+00
epoch 006/050 | total=8.361095 | rec=

/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/content/SpaMGCL/SpaMGCL/experiments/run_exp.py:101: UserWarning: Sparse invariant checks are implicitly disabled. Memory errors (e.g. SEGFAULT) will occur when operating on a sparse tensor which violates the invariants, but checks incur performance overhead. To silence this warning, explicitly opt in or out. See `torch.sparse.check_sparse_tensor_invariants.__doc__` for guidance.  (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:760.)
  return torch.sparse_coo_tensor(
